File: 253_Meeting_Rooms_II.py
# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        starts = []
        ends = []
        for interval in intervals:
            starts.append(interval.start)
            ends.append(interval.end)
        starts = sorted(starts)
        ends = sorted(ends)
        cur_earliest_end_int = 0
        total_req = 0
        for i in range(len(starts)):
            if starts[i] < ends[cur_earliest_end_int]:
                total_req += 1
            else:
                cur_earliest_end_int += 1
        return total_req

# Definition for an interval.
# class Interval:
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e

class Solution:
    def minMeetingRooms(self, intervals):
        """
        :type intervals: List[Interval]
        :rtype: int
        """
        intervals = sorted(intervals, key=lambda x: x.start)
        allocated = []
        for interval in intervals:
            if allocated and allocated[0] <= interval.start:
                heapq.heapreplace(allocated, interval.end)
            else:
                heapq.heappush(allocated, interval.end)
        return len(allocated)

File: 253_Meeting_Rooms_II.cpp
/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        vector<int> starts, ends;
        for (auto interval : intervals) {
            starts.push_back(interval.start);
            ends.push_back(interval.end);
        }
        sort(starts.begin(), starts.end());
        sort(ends.begin(), ends.end());
        int need = 0;
        for (int i = 0, cur = 0; i < starts.size(); ++i) {
            if (starts[i] < ends[cur])
                ++need;
            else
                ++cur;
        }
        return need;
    }
};

/**
 * Definition for an interval.
 * struct Interval {
 *     int start;
 *     int end;
 *     Interval() : start(0), end(0) {}
 *     Interval(int s, int e) : start(s), end(e) {}
 * };
 */
class Solution {
public:
    int minMeetingRooms(vector<Interval>& intervals) {
        sort(intervals.begin(), intervals.end(), [] (Interval a, Interval b) {return a.start < b.start;});
        priority_queue<int, vector<int>, greater<int>> needs;
        for (auto interval : intervals) {
            if (needs.empty() || interval.start < needs.top())
                needs.push(interval.end);
            else {
                needs.pop();
                needs.push(interval.end);
            }
        }
        return needs.size();
    }
};
